### ISSUE TITLE
File downloaded file gets truncated in linux if calling quitAndInstall.

### ISSUE THREAD

**Issue by manast** [megahertz/electron-simple-updater#89]
I am trying to implement automatic updates, so when the newer file is downloaded I want to quitAndInstall in order to use the new version. However the problem is that when I call ```quitAndInstall``` from the ```update-downloaded``` listener handler, the file gets truncated to around half the size (the file is around 100Mb, and it ends in between 50Mb and 40Mb, it is not deterministic. Makes me think that when the ```update-downladed``` event is triggered, the file has not been completely flushed to disk, so if you quit too quickly you end with a truncated file. If I do not call ```quitAndInstall``` then the file is downloaded correctly and if I manually quit and start the new version will work.

Coincidentally this made me wonder, what would happen if the download fails before the new version has been completely downloaded? currently it is overwriting the older version, so if it fails it seems like you end up with a broken app. This is normally resolved by saving to a new name and then doing a rename when ready, however I do not know if it is possible to do an atomic remove (old version) and rename (new version) so that it is guaranteed that you will broke an existing app.

Any insights ?

**Comment by manast** [comment]
I as I see at least in linux it is possible to rename a temporary file to an existing file to effectively achieve an atomic replacement of the old version by the new version: https://unix.stackexchange.com/questions/24395/rewrite-existing-file-so-that-it-gets-replaced-by-new-version-atomically-only-o

**Comment by manast** [comment]
Ok, I could figure out that currently the new version is being downloaded to the /tmp dir and then at the next quit it renames the new version to the old one and we are in business.
Regarding the issue with truncation, a workaround that seems to work is this:
```ts
  updater.on("update-downloaded", (meta: {}) => {
    logger.debug("update downloaded", meta);
    setTimeout(() => {
      updater.quitAndInstall();
    }, 5000);
  });
``` 
But it seems a bit shaky, and it should not be needed in the first place I think.

**Comment by manast** [comment]
Update. Seems like the workaround above does not work always, as if 5 seconds it is not enough for the whole file to be downloaded, it almost seems as if the event ```update-downladed``` is actually triggered *before* the whole file is downloaded :/

**Comment by megahertz** [comment]
Thanks for the information, I'll review again the downloader code on this week.

**Comment by manast** [comment]
Actually I think the issue comes from the httpreq package: 
https://github.com/SamDecrock/node-httpreq/blob/c2dbc959ea8538ffcbb12c5a7e7e1b90abd7d5e0/lib/httpreq.js#L330
It is listening to ```res.on('end')``` to decide the file has been downloaded and stored to disk, however it should be listening to the "target" stream instead, downloadStream.on('end'). Examining a bit more the quality of the httpreq code and the pretty low stars rate I would seriously consider to replace this library by something more battle tested and popular such as "request" or "node-fetch".

**Comment by manast** [comment]
in fact it should not even be listening to the "end" event but to the "finish" event. So a working code for download a file would be more in this style:
```ts
async function download() {
  const res = await fetch(srcFileUrl)
  await new Promise((resolve, reject) => {
    const fileStream = fs.createWriteStream('./octocat.png');
    res.body.pipe(fileStream);
    res.body.on("error", (err) => {
      reject(err);
    });
    fileStream.on("finish", function() {
      resolve();
    });
  });
}
```

**Comment by manast** [comment]
https://stackoverflow.com/questions/28334610/whats-the-difference-between-end-and-finish-events-in-node-streams

**Comment by manast** [comment]
Actually I think the issue I reporting here is the same as this one: https://github.com/megahertz/electron-simple-updater/issues/58

The reason why the poster of #58 got it working is because he implemented downloadFile correctly.

**Comment by megahertz** [comment]
Thank you very much for details. Yes, I'm thinking about migrating to axios. Unfortunately, I have no time at this moment to dive into the issue, will try to do that in the beginning of the next week, maybe earlier.

**Comment by manast** [comment]
Since this issue was a show stopper for me I ended using the standard electron-updater which actually works really well. Just in case somebody else encounters this issue thats the workaround I found.

**Comment by megahertz** [comment]
Sorry, sometimes I absolutely have no time for opensource. Anyway, I still have plans to make a release on the week.

**Comment by megahertz** [comment]
FIxed in v2.0.11
